### PR TITLE
[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* #60679 [DataLoader] Add Example Only `fork` DataPipe
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* #60675 [DataLoader] Add simple `groupby` DataPipe
* #60672 [DataLoader] Make `batch` DataPipe sensitive to unbatch_level argument
* **#60671 [DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument**
* #60670 [DataLoader] Rename transform DataPipe to legacy_transform

Differential Revision: [D29461083](https://our.internmc.facebook.com/intern/diff/D29461083)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 1]
```diff
@@ -74,19 +76,29 @@ class ShuffleIterDataPipe(IterDataPipe[T_co]):
...
         self._buffer = []
+                
+    def buffer_replace(self, x):
+        idx = random.randint(0, self.buffer_size - 1)
+        val = self._buffer[idx]
+        self._buffer[idx] = x
+        return val
 
     def __iter__(self) -> Iterator[T_co]:
+        # TODO: Buffer is global, should be per __iter__ !!!
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/combinatorics.py:98]: Is this TODO aiming to make shuffle shardable?

**Review by ejguan — APPROVED** [review]
LGTM, one question as below.

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/combinatorics.py:98]: It is just standard behaviour of iterators to be independent from each other

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).

**Pull request closed without merging**